pipe_gaps/pipeline/beam/transforms/core.py

```diff
@@ -58,7 +58,10 @@ def join_outputs(self):
     def _process_boundaries(self):
         side_inputs = None
         if self._side_inputs is not None:
-            side_inputs = beam.pvalue.AsList(self._side_inputs)
+            # side_inputs = beam.pvalue.AsList(self._side_inputs)
+            side_inputs = beam.pvalue.AsMultiMap(
+                self._side_inputs | beam.GroupBy(self._process.boundaries_key().func())
+            )
```

Comment:
I still think `CoGroupByKey` would be clearer in the general case, but for gaps particularly this should work well (I didn't know about `AsMultiMap`, pretty cool tool to have), so if you prefer we could leave this like you did and we then work on the `CoGroupByKey` approach when we implement something that needs that.